### DISCUSSION TITLE
Help with the new custom ESlint rule

### DISCUSSION THREAD

**Discussion by allurisriakhil** [eslint/eslint#17702] [Q&A] [answered]
Hi everyone,

I am new to ESlint, i am trying to learn and author a new custom rule using ESlint below are the steps i am following to create but it seems to be not working as expected facing some difficulty in debugging the issue. Please help me with the same. Thanks.

Custom Rule Def: I am authoring a new rule to identify the filenames in a directory and check wether the filename start with AK_ or not. If the file name doesn't start with AK_ then the rule should throw the error

Steps to recreate:

Step1: Created a new directory

`eslint-plugin-ak-rule`

Step2: Performed below command in the empty directory

`npm init`

Step3: Created the custom rule file with name 'ak-rule.js' in the same directory

Step4: The custom rule code is as below

```
module.exports = {
    meta: {
       type: 'suggestion',
       docs: {
         description: 'Enforce file names to start with "AK_"',
         category: 'Best Practices',
         recommended: true,
       },
    },
    create(context) {
        return {
            Program(node) {
              const fileName = context.getFilename();
              if (!fileName.startsWith('AK_')) {
                context.report({
                  node: node,
                  message: 'File names should start with "AK_".',
                });
              }
            },
        };
    },
};
```

Step5: My package.json file look as below

```
{
  "name": "eslint-plugin-ak-rule",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "eslintConfig": {
    "rules": {
      "ak-rule": "error"
    },
  "keywords": [],
  "author": "",
  "license": "ISC"
  }
}
```

Step6: Then i used command `npm link` to test the rule locally

Step7: Then in one of the project directories i created a sample file with name that doesn't start with AK_  to test

Step8: In that project directory i created a .eslintrc.json file as below

```
{
    "env": {
        "browser": true,
        "es2021": true
    },
    "extends": "eslint:recommended",
    "parserOptions": {
        "ecmaVersion": 6,
        "sourceType": "module"
    },
    "plugins": [
        "ak-rule"
    ],
    "rules": {
        "ak-rule/ak-rule": "error"
    }
}
```
Step9: In the same project directory i ran the below command 

`npm link eslint-plugin-ak-rule`

But the above rule seems to be not working while testing that in a the project directory locally not receiving any output. 

Please guide me with this rule if incase i miss any step or any of the above rule configurations are wrong. Thanks for your time in advance.

**Comment by Rec0iL99** [comment, marked as answer]
Hi @allurisriakhil, you need to provide a starting point for your plugin, where you would include your custom rules in the `rules` object.

https://eslint.org/docs/latest/extend/custom-rule-tutorial#step-7-bundle-the-custom-rule-in-a-plugin

**Reply by allurisriakhil** [reply]
Hi @Rec0iL99, 

Thanks for responding i tried the above suggestion by creating an entry point to the plugin as below

*** This seems to be giving error in both cases as shown below which shouldn't be the case could you please let me know if there is any error with my rule or files? ***

Step1: Created a new file in the same directory

eslint-plugin-ak.js

```
// eslint-plugin-ak.js

const akRule = require("./ak-rule");
const plugin = { rules: { "ak-rule": akRule } };
module.exports = plugin;
```

Step2: Then i followed the step 8 from the same link you provided and created eslint.config.js file

```
// eslint.config.js
"use strict";

// Import the ESLint plugin locally
const eslintPluginExample = require("./eslint-plugin-ak");

module.exports = [
    {
        files: ["**/*.js"],
        languageOptions: {
            sourceType: "commonjs",
            ecmaVersion: "latest",
        },
        // Using the eslint-plugin-example plugin defined locally
        plugins: {"example": eslintPluginExample},
        rules: {
            "example/ak-rule": "error",
        },
    }
]
```

Step3: Then tried to execute the rule by creating two sample files `sample.js` and `AK_sample.js` and tested using the commands below

```
npx eslint sample.js

/Users/<>/Desktop/eslint-plugin-ak-rule/sample.js
  2:1  error  File names should start with "AK_"  example/ak-rule

✖ 1 problem (1 error, 0 warnings)

npx eslint AK_sample.js

/Users/<>/Desktop/eslint-plugin-ak-rule/AK_sample.js
  1:1  error  File names should start with "AK_"  example/ak-rule

✖ 1 problem (1 error, 0 warnings)
```

**Reply by Rec0iL99** [reply]
That's because `context.getFilename()` returns the complete absolute path (for eg. `/Users/foo/test/hello.js`) and doing `.startsWith()` on it won't work as expected. You can use `path.basename()` to get the base file name.

```js
Program(node) {
  const fileName = path.basename(context.getFilename());
  if (!fileName.startsWith('AK_')) {
    context.report({
      node: node,
      message: 'File names should start with "AK_".',
     });
   }
},
```

Note: `context.getFilename()` is deprecated please use [`context.filename`](https://eslint.org/docs/latest/extend/custom-rules#the-context-object) instead.
